### ISSUE TITLE
Clarify installation steps

### ISSUE THREAD

**Issue by cbarrete** [wineasio/wineasio#70]
I've previously successfully setup/used wineasio, but I've just spent a few hours unsuccessfully trying to get it to work on a different system.
It feels to me that the main reason I'm having trouble is that it's unclear what is currently the best way to install wineasio. The instructions in the README clash with what users report is working for them in issues and on other forums, new versions of wine seem to behave differently, etc.

A few things that I wish were clarified include:
- Is it required to install the dll/so files in `/usr` (or the equivalent on the system)? What about rootless installs (e.g. with package managers such as Nix or Conda), are they not supported?
- Should the dll/so files be copied in the wine prefix that is going to use them? I've also seen some people recommend copying them next to the binary (`.exe`) that is going to load them.
- What should the argument to `regsvr32`/`wine64 regsvr32` be? Should it be an absolute path? If the files have been installed in various places (system, wine prefix, etc.), which one(s) should be registered?
- Should the .so files also be registered, or only the dlls?
- Is `WINEDLLPATH` relevant to the installation process? If so, how?

I've tried various combinations of the following, but wineasio is either not detected by the application (FL Studio) or I get an error when trying to use it ("Couldn't open the ASIO driver").

Any clarifications would be appreciated, and I'd be happy to submit a PR to update the README to help other users.

Thanks!

**Comment by williamcorney** [comment]
There is a package from kxstudio which you can probably use depending on your distribution of linux ( which is what I assume you are using)

> * Is it required to install the dll/so files in `/usr` (or the equivalent on the system)? What about rootless installs (e.g. with package managers such as Nix or Conda), are they not supported?

Yes install files wineasio.dll and wineasio.dll.so into /lib/wine/x86_64-windows and  lib/wine/x86_64-unix respectively.  I haven't any experience of rootless installs so can't comment here.

> * Should the dll/so files be copied in the wine prefix that is going to use them? I've also seen some people recommend copying them next to the binary (`.exe`) that is going to load them.

Files don't need to be copied into the wine prefix (eg into the c:/windows/system32 folder).
the regsvr32 command needs to be run in any prefix you are using but it looks for the files in /lib/wine/x86_64-windows and  lib/wine/x86_64-unix

If you are just using the default system wine then you need to copy the files to one place and register in the prefix you want to use.

If you are using a third party wine, be mindful that it may have its own copy of the wine binary and a replica of the above folders.  If this is the case you may have to register wineasio.dll first with the system wine ( but specifying WINEPREFIX="/blabla/bla/locationofmywinebottle" ) and then register wineasio.dll with that bottlers wine cmd prompt.

> * What should the argument to `regsvr32`/`wine64 regsvr32` be? Should it be an absolute path? If the files have been installed in various places (system, wine prefix, etc.), which one(s) should be registered?

Usually just regsvr32 wineasio.dll will do the trick.

> * Should the .so files also be registered, or only the dlls?

Only the dll's - though in the background I think its actually the dll.so file where the magic happens.

> * Is `WINEDLLPATH` relevant to the installation process? If so, how?

Not in my experience Inasmuch I haven't had to change or do anything with that.

> 
> I've tried various combinations of the following, but wineasio is either not detected by the application (FL Studio) or I get an error when trying to use it ("Couldn't open the ASIO driver").

You need in my experience to have jack dump installed from https://jackaudio.org.  My experience is more with OSX than Linux 
> 
> Any clarifications would be appreciated, and I'd be happy to submit a PR to update the README to help other users.
> 
> Thanks!

**Comment by cbarrete** [comment]
Thanks for the detailed answer!

> There is a package from kxstudio which you can probably use depending on your distribution of linux ( which is what I assume you are using)

No actually, I'm using the [Nixpkgs wineasio package](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/emulators/wineasio/default.nix), which just builds wineasio from source and installs it somewhere in `/nix/store/`.
When I run `wine64 regsvr32 /nix/store/r07086knaph2yk75vi7dgzh9f0w14y7y-wineasio-1.1.0/lib/wine/x86_64-windows/wineasio.dll`, it fails to load the DLL. If I fixup `WINEDLLPATH`, it does get loaded properly, but it's still not recognized by the audio application.

> Yes install files wineasio.dll and wineasio.dll.so into /lib/wine/x86_64-windows and lib/wine/x86_64-unix respectively

That's not an option in my case, since this would require root access (and in the case of NixOS specifically, `/lib` doesn't exist anyway).

> the regsvr32 command needs to be run in any prefix you are using but it looks for the files in /lib/wine/x86_64-windows and lib/wine/x86_64-unix

I guess that's what I'll have to look into then, getting `regsvr32` to look elsewhere, or accept my absolute paths without WINEDLLPATH magic.

**Comment by techninja1008** [comment]
@cbarrete I am also currently trying to get this to work on NixOS - did you succeed?

**Comment by cbarrete** [comment]
@techninja1008 Yes, here is a sample home-manager module that adds a `register-wineasio` script to your environment, which you just have to call once to setup wineasio.
You could easily adapt it to a NixOS module instead, or just inline it e.g. in your `environment.systemPackages`.

```nix
{ pkgs, ... }:
let
  winePkg = pkgs.wineWowPackages.stable;
  wine = "${winePkg}/bin/wine64";
  so = "${pkgs.wineasio}/lib/wine/x86_64-unix/wineasio64.dll.so";
  dest = "$WINEPREFIX/drive_c/windows/system32/wineasio64.dll";
in {
  home.packages = [
    (pkgs.writeShellScriptBin "register-wineasio" ''
      cp -v ${so} ${dest}
      chmod +w ${dest}
      ${wine} regsvr32 ${so}
    '')
  ];
}
```

I haven't had issues with it so far on a non-NixOS system that I manage with Nix.
Note that it has different logic from the upstream `register-wineasio` script in that it only sets up the `.dll.so` file, renamed to `.dll`.
I might try to upstream it to `nixpkgs` later if it works for you too.

**Comment by techninja1008** [comment]
@cbarrete I've just tried this with Ableton Live 11 Suite - it results in Live crashing as soon as it tries to list ASIO drivers. Using the upstream script allows the driver to be listed but gives an error when attempting to load it - not a crash though.

**Comment by TotalChaos05** [comment]
@techninja1008 @cbarrete I'm having the same issue with ableton live 11 and nixos, are there any known workarounds

**Comment by cbarrete** [comment]
This used to work on a Debian + Nix workstation, now that I've switched it to NixOS, it has also stopped working for me and FL Studio just hangs when selecting the WineASIO driver... I'll let you know if I get it working again.

**Comment by TotalChaos05** [comment]
I have a messy looking workaround where i switch wine to using alsa, and copy wineasio64.dll.so to windows/system/wineasio(64).dll and windows/system32/wineasio(64).dll
